### DISCUSSION TITLE
How can I create a KTX2 texture from buffer that is in memory and save it to a file?

### DISCUSSION THREAD

**Discussion by GasimGasimzada** [KhronosGroup/KTX-Software#554] [Q&A] [answered]
I can't figure out how to use the KTX writer module or just manually create a KTX2 texture. I am looking at the following API reference: https://www.khronos.org/ktx/documentation/libktx/group__writer.html

Firstly, I do not have access to `ktxWriteKTXF`. I checked the include file in this repo as well. The function does not exist. So, I tried to use the [create function](https://www.khronos.org/ktx/documentation/libktx/group__writer.html#gaa051e6b5fe8d46c4af43cb03607c4b37):

```cpp
ktxTexture *tex = nullptr;

// This function does not exist. Only ktxTexture2_Create exists
// but there are no counterparts for ktxTexture_SetImageFromMemory that
// allows using ktxTexture2
auto res = ktxTexture_Create(&createInfo, KTX_TEXTURE_CREATE_ALLOC_STORAGE,
                               &ktxTexture);

ktxTexture_SetImageFromMemory(
    ktxTexture, 1, 1, 1, static_cast<const ktx_uint8_t *>(textureDataLoadedFromSTB),
    textureSize);

FILE *file = fopen(fileName.string().c_str(), "wb");
ktxTexture_WriteToStdioStream(ktxTexture, file);
```

What am I missing here? How can I access the writer API?

I am using KTX4.0.0 and building it with the following Cmake flags:

```json
"KTX_FEATURE_TESTS": "OFF",
"KTX_FEATURE_TOOLS": "OFF",
"KTX_FEATURE_VULKAN": "OFF",
"KTX_FEATURE_KTX1": "OFF",
"KTX_FEATURE_STATIC_LIRBARY": "ON",
```

**Comment by MarkCallow** [comment, marked as answer]
Somehow you have ended up at a very old version of the documentation. ktxWriteKTXF was deprecated a very long time and subsequently removed. Here is the [latest](https://github.khronos.org/KTX-Software/libktx/group__writer.html) version of the page you were looking at. See also the [Introduction](https://github.khronos.org/KTX-Software/libktx/index.html) of the libktx documentation for a good summary of how to do various things with the library.

How did you end up at that page? I don't see a link to it on the KTX landing page at www.khronos.org/ktx or any of the pages linked from there. The only link I found points to the documentation I just pointed you to.

I'll ask the Khronos Webmaster to update https://www.khronos.org/ktx/documentation.

**Reply by GasimGasimzada** [reply]
That's my bad and it makes sense now 😄  I just googled the docs and it took me to this page. The URL pointed me to this the khornos domain; so, I thought it was official one.

**Reply by baiwenbo1997** [reply]
Hello, I am an initiator. I would like to convert the PNG format into KTX2 format, how should I operate? I refer to https://github.khronos.org/ktx-software/libktx/index.html#writektx, it should be that the way does not have the created KTX2 is not correct, do you have a more detailed example? thank you very much

**Reply by GasimGasimzada** [reply]
In that page, the section titled "Writing a KTX or KTX2 file" worked for me for KTX2 file.

**Comment by MarkCallow** [comment]
@baiwenbo1997 the `ktxTexture2_Write*` functions write a ktxTexture2 object in memory to a file. That is they serialize existing KTX2 data. They do not do any conversion. Use `toktx` to create KTX2 files from PNG files. If you want to write your own software to do this, also look at `toktx` to see how to do it.
